File: templates/python/code-carl/ch06_tree/t13_lc0404_sum_of_left_leaves.py
"""LeetCode 404. Sum of Left Leaves
https://leetcode.cn/problems/sum-of-left-leaves/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        pass


if __name__ == "__main__":
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    solution = Solution()
    assert solution.sumOfLeftLeaves(root) == 24
    solution = Solution()
    assert solution.sumOfLeftLeaves(None) == 0
    print("All tests passed.")

File: solutions/code-carl/ch06_tree/T13_LC0404_SumOfLeftLeaves.java
/**
 * Practice template for:
 * - LeetCode 404. Sum of Left Leaves
 *   https://leetcode.cn/problems/sum-of-left-leaves/
 */
public class T13_LC0404_SumOfLeftLeaves {
    /*    public int sumOfLeftLeaves(TreeNode root) {
            if (root == null) {
                return 0;
            }

            int leftSum = 0, rightSum = 0;
            if (root.left != null && root.left.left == null && root.left.right == null) {
                leftSum = root.left.val;
            } else {
                leftSum = sumOfLeftLeaves(root.left);
            }
            rightSum = sumOfLeftLeaves(root.right);
            return leftSum + rightSum;
        }*/
    public int sumOfLeftLeaves(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left;
        if (root.left != null && root.left.left == null && root.left.right == null) {
            left = root.left.val;
        } else {
            left = sumOfLeftLeaves(root.left);
        }
        int right = sumOfLeftLeaves(root.right);
        return left + right;
    }
}
